File: blackjack/cpp/outcomes.cpp
#include <iostream>
using namespace std;

int partitions(int cards[10], int subtotal) {
  int m=0;
  int total;
  // Hit
  for (int i = 0; i < 10; i++) {
    if (cards[i]>0) {
      total = subtotal+i+1;
      if (total < 21) {
	// Stand
	m += 1;
	// Hit again
	cards[i] -= 1;
	m += partitions(cards, total);
	cards[i] += 1;
      } else if (total==21) {
	// Stand; hit again is an automatic bust
	m += 1;
	break;
      }
    }
  }
  return m;
}

int main(void) {
  int deck[] = {4,4,4,4,4,4,4,4,4,16};
  int d=0;
  
  for (int i = 0; i < 10; i++) {
    // Dealer showing
    deck[i] -= 1;
    int p = 0;
    for (int j = 0; j < 10; j++) {
      deck[j] -= 1;
      p += partitions(deck, j+1);
      deck[j] += 1;
    }

    cout << "Dealer showing " << i << " partitions = " << p << "\n";
    d += p;
    deck[i] += 1;
  }
  cout << "Total partitions = " << d << "\n";

  return 0;
}

File: blackjack/lua/outcomes.lua
function partitions (cards, subtotal)
  local result = 0

  -- Hit
  for i = 1,10 do
    if (cards[i]>0) then
      local total = subtotal+i
      if (total < 21) then
        -- Stand
        result = result+1
        -- Hit again
        cards[i] = cards[i]-1
        result = result+partitions(cards, total)
        cards[i] = cards[i]+1
      elseif (total==21) then
        -- Stand; hit again is an automatic bust
        result = result+1
        break
      end
    end
  end
  return result
end

deck = {4,4,4,4,4,4,4,4,4,16}

d = 0

for i = 1,10 do
    -- Dealer showing
    deck[i] = deck[i]-1
    p = 0
    for j = 1,10 do
        deck[j] = deck[j]-1
        p = p+partitions(deck, j)
	deck[j] = deck[j]+1
    end
    io.write("Dealer showing ",i-1," partitions = ",p,"\n")
    d = d+p
    deck[i] = deck[i]+1
end

io.write("Total partitions = ",d,"\n")
